File: app/src/main/java/com/feresr/imdino/MainActivity.java
package com.feresr.imdino;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.webkit.JavascriptInterface;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;

public class MainActivity extends AppCompatActivity {

    public static final int DINO_COUNT = 5;
    private TextView speedTextView;
    private TextView distanceTextView;
    private TextView resultTextView;
    private TextView generationTextView;
    private WebView webView;
    private KeyEvent keyEvent;
    private int maxScoreInGeneration = 0;
    private int minScoreInGeneration = Integer.MAX_VALUE;
    private int generation = 0;
    private int currentDino = 0;
    private Dino bestDino;
    private Handler handler;
    private LinearLayout dinoPlaceholderView;

    private ArrayList<Dino> dinosaurs;

    @SuppressLint({"SetJavaScriptEnabled", "JavascriptInterface", "AddJavascriptInterface"})
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        webView = (WebView) findViewById(R.id.webview);
        final WebSettings webSettings = webView.getSettings();
        webSettings.setJavaScriptEnabled(true);
        webView.addJavascriptInterface(this, "Android");
        webView.loadUrl("file:///android_asset/index.html");

        speedTextView = (TextView) findViewById(R.id.speed);
        distanceTextView = (TextView) findViewById(R.id.distance);
        resultTextView = (TextView) findViewById(R.id.result);
        generationTextView = (TextView) findViewById(R.id.generation);

        keyEvent = new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_SPACE);

        handler = new Handler();

        dinosaurs = new ArrayList<>();
        LayoutInflater layoutInflater = LayoutInflater.from(this);
        dinoPlaceholderView = (LinearLayout) findViewById(R.id.dino_placeholder);
        for (int i = 0; i < DINO_COUNT; i++) {
            dinosaurs.add(new Dino());
            layoutInflater.inflate(R.layout.dino_image, dinoPlaceholderView);
        }
    }

    @JavascriptInterface
    public void onGameUpdate(final float obstacleX, final float currentSpeed) {
        if (dinosaurs.get(currentDino).think(new float[]{obstacleX, currentSpeed})) {
            webView.dispatchKeyEvent(keyEvent);
        }
        handler.post(new Runnable() {
            @Override
            public void run() {
                speedTextView.setText(String.valueOf(currentSpeed));
                distanceTextView.setText(String.valueOf(obstacleX));
                resultTextView.setText(String.valueOf(dinosaurs.get(currentDino).getThinkValue()));
            }
        });
    }

    @JavascriptInterface
    public void onGameStart() {
        Log.i("onGameStart", "Running dino: " + currentDino + " in generation " + generation + " " + Arrays.toString(dinosaurs.get(currentDino).getGenome()));
        handler.post(new Runnable() {
            @Override
            public void run() {
                if (currentDino == 0) {
                    dinoPlaceholderView.getChildAt(DINO_COUNT - 1).setAlpha(0.25f);
                } else {
                    dinoPlaceholderView.getChildAt(currentDino - 1).setAlpha(0.25f);
                }

                dinoPlaceholderView.getChildAt(currentDino).setAlpha(1.0f);
                generationTextView.setText("Generation: " + String.valueOf(generation) + " • Max score in gen: " + maxScoreInGeneration);
            }
        });
    }

    @JavascriptInterface
    public void onGameOver(int score) {
        Log.i("onGameOver", "Running dino: " + currentDino + " score: " + score);

        dinosaurs.get(currentDino).setDistanceRan(score);
        if (score > maxScoreInGeneration) {
            maxScoreInGeneration = score;
            bestDino = dinosaurs.get(currentDino);
        }

        if (score < minScoreInGeneration) {
            minScoreInGeneration = score;
        }

        if (currentDino < dinosaurs.size() - 1) {
            currentDino++;
            reloadGame();
        } else {
            toNextGeneration();
        }

    }

    private void reloadGame() {
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                webView.dispatchKeyEvent(new KeyEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_SPACE));
            }
        }, 1000);
    }


    private void toNextGeneration() {
        Log.i("toNextGeneration", "GENERATION " + generation + "FINISHED best dino: " + bestDino.getDistanceRan());

        Random r = new Random();
        ArrayList<Dino> survivers = new ArrayList<>();
        for (int i = 0; i < DINO_COUNT; i++) {
            if (r.nextInt(maxScoreInGeneration - minScoreInGeneration) < dinosaurs.get(i).getDistanceRan() - minScoreInGeneration) {
                survivers.add(dinosaurs.get(i));
            } else {
                Log.i("Dino", "dino killed: " + dinosaurs.get(i).getDistanceRan());
            }
        }

        for (int i = survivers.size(); i < DINO_COUNT; i++) {
            survivers.add(survivers.get(r.nextInt(survivers.size())).reproduce());
        }

        generation++;
        currentDino = 0;
        maxScoreInGeneration = 0;
        minScoreInGeneration = Integer.MAX_VALUE;
        this.dinosaurs = survivers;
        reloadGame();
    }
}

File: app/src/main/java/com/feresr/imdino/MainActivity.kt
package com.feresr.imdino

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.KeyEvent
import android.view.LayoutInflater
import android.webkit.JavascriptInterface
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

class MainActivity : AppCompatActivity() {

    private val keyEvent: KeyEvent = KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_SPACE)
    private val handler: Handler = Handler()
    private var maxScoreInGeneration = 0
    private var generation = 0
    private var currentDino = 0
    private var bestDino: Dino? = null
    private var jumps:Int = 0
    private val JUMP_PENALTY:Int = 100
    private var dinosaurs: ArrayList<Dino> = ArrayList()

    @SuppressLint("SetJavaScriptEnabled", "JavascriptInterface", "AddJavascriptInterface")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        webview.settings.javaScriptEnabled = true
        webview.addJavascriptInterface(this, "Android")
        webview.loadUrl("file:///android_asset/index.html")

        val layoutInflater = LayoutInflater.from(this)
        for (i in 0..DINO_COUNT - 1) {
            dinosaurs.add(Dino())
            layoutInflater.inflate(R.layout.dino_image, dino_placeholder)
        }
    }

    @JavascriptInterface
    fun onJump() {
        jumps++ // keep track of jumps to punish jumpers
    }

    @JavascriptInterface
    fun onGameUpdate(obstacleX: Float, currentSpeed: Float) {
        if (dinosaurs[currentDino].think(floatArrayOf(obstacleX, currentSpeed))) {
            webview.dispatchKeyEvent(keyEvent)
        }
        handler.post {
            speedTxtView.text = currentSpeed.toString()
            distanceTxtView.text = obstacleX.toString()
            if (currentDino > 0 && currentDino < dinosaurs.size) {
                resultTxtView.text = dinosaurs[currentDino].thinkValue.toString()
            }
        }
    }

    @JavascriptInterface
    fun onGameStart() {
        Log.i("onGameStart", "Running dino: " + currentDino + " in generation " + generation + " " + Arrays.toString(dinosaurs[currentDino].genome))
        handler.post {
            if (currentDino == 0) {
                dino_placeholder.getChildAt(DINO_COUNT - 1).alpha = 0.25f
            } else {
                dino_placeholder.getChildAt(currentDino - 1).alpha = 0.25f
            }

            dino_placeholder.getChildAt(currentDino).alpha = 1.0f
            generationTxtView.text = "Generation: $generation • Max score in gen: $maxScoreInGeneration"
        }
    }

    @JavascriptInterface
    fun onGameOver(score: Int) {
        Log.i("onGameOver", "Running dino: $currentDino score: $score")

        dinosaurs[currentDino].distanceRan = score - (jumps * JUMP_PENALTY)
        jumps = 0
        if (score > maxScoreInGeneration) {
            maxScoreInGeneration = score
            bestDino = dinosaurs[currentDino]
        }

        if (currentDino < dinosaurs.size - 1) {
            currentDino++
            reloadGame()
        } else {
            toNextGeneration()
        }
    }

    private fun reloadGame() {
        handler.postDelayed({ webview.dispatchKeyEvent(KeyEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_SPACE)) }, 1000)
    }

    private fun toNextGeneration() {
        Log.i("toNextGeneration", "GENERATION " + generation + "FINISHED best dino: " + bestDino?.distanceRan)

        val r = Random()
        val survivors = ArrayList<Dino>()

        val w = DoubleArray(DINO_COUNT)

        var best = 0
        dinosaurs.asSequence()
                .filter { it.distanceRan > best }
                .forEach { best = it.distanceRan }

        var wmax = 0.0
        for (i in 0..DINO_COUNT - 1) {
            w[i] = Utils.pdf(dinosaurs[i].distanceRan.toDouble(), best.toDouble(), 2.0)
            if (w[i] > wmax) {
                wmax = w[i]
            }
        }

        var b = 0.0

        var index = r.nextInt(DINO_COUNT)
        for (i in 0..DINO_COUNT - 1) {
            b += r.nextDouble() * 2.0 * wmax

            while (w[index % w.size] <= b) {
                b -= w[index % w.size]
                index += 1
            }
            survivors.add(dinosaurs[index % dinosaurs.size])
        }

        this.dinosaurs.clear()

        for (i in 0..DINO_COUNT - 1) {
            dinosaurs.add(survivors[i].reproduce())
        }
        generation++
        currentDino = 0
        maxScoreInGeneration = 0
        reloadGame()
    }

    companion object {
        val DINO_COUNT = 6
    }
}
